{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "d:\\VirtualEnvironment\\Anaconda3\\envs\\TF-krease\\lib\\site-packages\\numpy\\_distributor_init.py:30: UserWarning: loaded more than 1 DLL from .libs:\n",
      "d:\\VirtualEnvironment\\Anaconda3\\envs\\TF-krease\\lib\\site-packages\\numpy\\.libs\\libopenblas.EL2C6PLE4ZYW3ECEVIV3OXXGRN2NRFM2.gfortran-win_amd64.dll\n",
      "d:\\VirtualEnvironment\\Anaconda3\\envs\\TF-krease\\lib\\site-packages\\numpy\\.libs\\libopenblas.FB5AE2TYXYH2IJRDKGDGQ3XBKLKTF43H.gfortran-win_amd64.dll\n",
      "d:\\VirtualEnvironment\\Anaconda3\\envs\\TF-krease\\lib\\site-packages\\numpy\\.libs\\libopenblas.GK7GX5KEQ4F6UYO3P26ULGBQYHGQO7J4.gfortran-win_amd64.dll\n",
      "d:\\VirtualEnvironment\\Anaconda3\\envs\\TF-krease\\lib\\site-packages\\numpy\\.libs\\libopenblas64__v0.3.21-gcc_10_3_0.dll\n",
      "  warnings.warn(\"loaded more than 1 DLL from .libs:\"\n"
     ]
    },
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'utils'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[1], line 10\u001b[0m\n\u001b[0;32m      8\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mmatplotlib\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpyplot\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mplt\u001b[39;00m\n\u001b[0;32m      9\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01msys\u001b[39;00m\n\u001b[1;32m---> 10\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mutils\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mCLR\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;241m*\u001b[39m\n\u001b[0;32m     11\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mutils\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mLoadData\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m load_CW_Source\n\u001b[0;32m     12\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mModel\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mCNN\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m cnn_classifier\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'utils'"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "parent_dir = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "import tensorflow as tf\n",
    "from sklearn import preprocessing\n",
    "import numpy as np\n",
    "import random\n",
    "import matplotlib.pyplot as plt\n",
    "import sys\n",
    "from utils.CLR import*\n",
    "from utils.LoadData import load_CW_Source\n",
    "from Model.CNN import cnn_classifier\n",
    "from sklearn.utils import shuffle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tensorflow.keras.callbacks import ModelCheckpoint\n",
    "from tensorflow.keras.utils import to_categorical\n",
    "def train_model(X_profiling, Y_profiling, X_test, Y_test, model, save_file_name, epochs=150, batch_size=100,\n",
    "                max_lr=1e-3):\n",
    "\n",
    "    # Save model every epoch\n",
    "    save_model = ModelCheckpoint(\n",
    "                    filepath=save_file_name,\n",
    "                    monitor='val_loss',\n",
    "                    mode='min',\n",
    "                    save_best_only=True)\n",
    "\n",
    "    # Get the input layer shape\n",
    "    input_layer_shape = model.get_layer(index=0).input_shape[0]\n",
    "\n",
    "    # Sanity check\n",
    "    if input_layer_shape[1] != len(X_profiling[0]):\n",
    "        print(\"Error: model input shape %d instead of %d is not expected ...\" % (\n",
    "        input_layer_shape[1], len(X_profiling[0])))\n",
    "        sys.exit(-1)\n",
    "    Reshaped_X_profiling, Reshaped_X_test = X_profiling.reshape(\n",
    "        (X_profiling.shape[0], X_profiling.shape[1], 1)), X_test.reshape((X_test.shape[0], X_test.shape[1], 1))\n",
    "\n",
    "    # One Cycle Policy\n",
    "    lr_manager = OneCycleLR(max_lr=max_lr, end_percentage=0.2, scale_percentage=0.1, maximum_momentum=None,\n",
    "                            minimum_momentum=None, verbose=True)\n",
    "\n",
    "    callbacks = [save_model, lr_manager]\n",
    "\n",
    "    history = model.fit(x=Reshaped_X_profiling, y=to_categorical(Y_profiling, num_classes=256),\n",
    "                        validation_data=(Reshaped_X_test, to_categorical(Y_test, num_classes=256)),\n",
    "                        batch_size=batch_size, verbose=1, epochs=epochs, callbacks=callbacks)\n",
    "    return history"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b> Trained on the dataset AES_device2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#region Configuration\n",
    "TRAINING_CONFIG = {\n",
    "    'bytes_to_train': range(0, 16),        # Process all 16 AES state bytes\n",
    "    'input_size': 800,                    # Number of features per trace\n",
    "    'learning_rate': 5e-3,                # Initial learning rate\n",
    "    'batch_size': 512,                      # Training batch size\n",
    "    'epochs': 100,                         # Training epochs        \n",
    "    'save_file_name': '../Model/', # Model save location\n",
    "}\n",
    "#endregion\n",
    "\n",
    "#region Main Training Loop\n",
    "for byte_idx in TRAINING_CONFIG['bytes_to_train']:\n",
    "    #region Session Management\n",
    "    tf.keras.backend.clear_session()  # Prevent model parameter leakage\n",
    "    #endregion\n",
    "\n",
    "    #region Data Loading\n",
    "    try:\n",
    "        # Load byte-specific dataset\n",
    "        (x_train, y_train, \n",
    "         x_test, y_test, \n",
    "         c_train, c_test) = load_CW_Source(in_file='../Dataset/AES_device2/',sec=18000,byte=byte_idx)\n",
    "        \n",
    "    except Exception as e:\n",
    "        print(f\"Data loading failed for byte {byte_idx}: {str(e)}\")\n",
    "        continue\n",
    "    #endregion\n",
    "\n",
    "\n",
    "    # Shuffle training data\n",
    "    x_train, y_train = shuffle(x_train, y_train)\n",
    "  \n",
    "\n",
    "\n",
    "    #region Data Preprocessing\n",
    "    # 1. Standardization\n",
    "    std_scaler = preprocessing.StandardScaler()\n",
    "    x_train = std_scaler.fit_transform(x_train.astype('float32'))\n",
    "    x_test = std_scaler.transform(x_test.astype('float32'))\n",
    "\n",
    "    # 2. Normalization\n",
    "    minmax_scaler = preprocessing.MinMaxScaler(feature_range=(0, 1))\n",
    "    x_train = minmax_scaler.fit_transform(x_train)\n",
    "    x_test= minmax_scaler.transform(x_test)\n",
    "\n",
    "    #endregion.\n",
    "\n",
    "\n",
    "    try:\n",
    "        # Model initialization\n",
    "        model = cnn_classifier(\n",
    "            input_size=TRAINING_CONFIG['input_size'],\n",
    "            learning_rate=TRAINING_CONFIG['learning_rate']\n",
    "        )\n",
    "        model.summary()\n",
    "        \n",
    "        # Training execution\n",
    "        print(f\"\\n############### Training Byte {byte_idx} #################\\n\")\n",
    "        history = train_model(\n",
    "            x_train, y_train,\n",
    "            x_test, y_test,\n",
    "            model=model,\n",
    "            save_file_name=f\"{TRAINING_CONFIG['save_file_name']}Source_Model_byte{byte_idx}_D1.h5\",\n",
    "            epochs=TRAINING_CONFIG['epochs'],\n",
    "            batch_size=TRAINING_CONFIG['batch_size'],\n",
    "            max_lr=TRAINING_CONFIG['learning_rate']\n",
    "        )\n",
    "        \n",
    "        print(f\"\\n############### Byte {byte_idx} Training Completed #################\\n\")\n",
    "        \n",
    "    except Exception as e:\n",
    "        print(f\"Training failed for byte {byte_idx}: {str(e)}\")\n",
    "        continue"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "TF-krease",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
